xdas/compose.py

```diff
@@ -0,0 +1,530 @@
+
+    def __str__(self) -> str:
+        args = [f"{key}={val}" if (len(str(val)) < 10) else f"{key}=..." for key, val in self._kwargs.items()]
+        argstr = ", ".join(args)
+        func_name = getattr(self._func, "__name__", "[custom]")
+        return f"{self._name:<25}{func_name}({argstr})"
+
+    def _locate_before(self, x, a):
+        return x.index(a)
+
+    def _locate_after(self, x, a):
+        return x.index(a) + 1
+
+    def delete(self) -> None:
+        self._parent.pop(self._name)
+        pass
+
+    def set_args(self, **kwargs) -> None:
+        self._kwargs = kwargs
+        pass
+
+    def insert_after(self, atom) -> None:
+        self._parent._insert(self._name, atom, locator=self._locate_after)
+        pass
+
+    def insert_before(self, atom) -> None:
+        self._parent._insert(self._name, atom, locator=self._locate_before)
+        pass
+
+    def move_down(self) -> None:
+        self._parent._move(self._name, locator=self._locate_after)
+        pass
+
+    def move_up(self) -> None:
+        self._parent._move(self._name, locator=self._locate_before)
+        pass
+
+
+class StateAtom(Atom):
+    """
+    A subclass of Atom that provides some logic for handling
+    data states, which need to be updated throughout the
+    execution chain. An example of a stateful operation is
+    a recursive filter, passing on the state from t to t+1.
+
+    The Atom base class assumes that a given function takes
+    an xarray Database as the first argument, and returns
+    a modified copy of this Database. The StateAtom class
+    assumes that the stateful function takes a Database and
+    an initialized state, and returns the modified Database
+    and modified state, i.e.:
+
+    >>> db, state = func(db, state_arg=state, **kwargs)
+
+    Here, `state_arg` is the name of the keyword argument
+    that contains the state, which can differ from one function
+    to the next. For example, in scipy.signal.sosfilt, the
+    state argument is `zi`, and so `state_arg` is `zi`.
+
+    Methods
+    -------
+    initialize_state(state):
+        Set the initial state.
+
+    Examples
+    --------
+    >>> state = np.zeros((2, 2, 100))
+    ... state_op = StateAtom(xp.sosfilt, dim="time", state_arg="zi", state=state)
+
+    """
+
+    _state_initialized = False
+
+    def __init__(
+        self, func: Callable, state_arg: Hashable, state: str | Any = "init", **kwargs
+    ) -> None:
+
+        self._state_arg = state_arg
+        self._state = state
+        if state != "init":
+            self._state_initialized = True
+
+        super().__init__(func, **kwargs)
+
+    def __str__(self) -> str:
+        return super().__str__() + "  [stateful]"
+
+    def __call__(self, db) -> Any:
+        kwargs = self._kwargs.copy()
+        kwargs.update({self._state_arg: self._state})
+        db, state = self._func(db, **kwargs)
+        self._set_state(state)
+        return db
+
+    def _set_state(self, state: Any) -> None:
```

Comment:
We might want to change how we handle states in the (near) future, so I reserved a spot for it in the namespace. I could imagine states that need to be written to disk, or handled differently in the case of e.g. LSTM models. The former would allow one to continue a processing chain the next day without resetting the state, or recover from unintended interruptions of the pipeline.

Reply:
Good point about being able to restore a state the following day.

Reply:
Marie will need this feature if she wants to process the data day-by-day